src/terminal.cc

```diff
@@ -0,0 +1,275 @@
+#include "terminal.h"
+#include <iostream>
+#include "logging.h"
+#include "singletons.h"
+#include "process.h"
+
+#include <iostream> //TODO: remove
+using namespace std; //TODO: remove
+
+Terminal::InProgress::InProgress(const std::string& start_msg): stop(false) {
+  waiting_print.connect([this](){
+    Singleton::terminal->async_print(line_nr-1, ".");
+  });
+  start(start_msg);
+}
+
+Terminal::InProgress::~InProgress() {
+  stop=true;
+  if(wait_thread.joinable())
+    wait_thread.join();
+}
+
+void Terminal::InProgress::start(const std::string& msg) {
+  line_nr=Singleton::terminal->print(msg+"...\n");
+  wait_thread=std::thread([this](){
+    size_t c=0;
+    while(!stop) {
+      if(c%100==0)
+        waiting_print();
+      std::this_thread::sleep_for(std::chrono::milliseconds(10));
+      c++;
+    }
+  });
+}
+
+void Terminal::InProgress::done(const std::string& msg) {
+  if(!stop) {
+    stop=true;
+    Singleton::terminal->async_print(line_nr-1, msg);
+  }
+}
+
+void Terminal::InProgress::cancel(const std::string& msg) {
+  if(!stop) {
+    stop=true;
+    Singleton::terminal->async_print(line_nr-1, msg);
+  }
+}
+
+Terminal::Terminal() {
+  bold_tag=get_buffer()->create_tag();
+  bold_tag->property_weight()=PANGO_WEIGHT_BOLD;
```

Comment:
Saw that this is texttag which doesn't have operator|() overloaded so I guess just ignore this.